fedimint-api/src/db/notifications.rs

```diff
@@ -0,0 +1,154 @@
+use std::{
+    collections::hash_map::DefaultHasher,
+    hash::{Hash, Hasher},
+};
+
+use bitvec::vec::BitVec;
+use tokio::sync::{futures::Notified, Notify};
+
+// arbitrary number. if you have performance problems, tweak this
+const NOTIFY_BUCKETS: usize = 32;
+
+/// Manages notifications for the database
+#[derive(Debug)]
+pub struct Notifications {
+    slots: Vec<Notify>,
+}
+
+impl Default for Notifications {
+    fn default() -> Self {
+        Self {
+            slots: (0..NOTIFY_BUCKETS).map(|_| Notify::new()).collect(),
+        }
+    }
+}
+
+impl Notifications {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Waits untils the key *might* have changed.
+    ///
+    /// This registers for notification when called.
+    /// Then waits for the notification when .awaited.
+    pub fn wait<K>(&self, key: &K) -> Notified
+    where
+        K: Hash,
+    {
+        let mut hasher = DefaultHasher::new();
+        key.hash(&mut hasher);
+        let hash_value = hasher.finish();
+        let slot_idx = (hash_value % (self.slots.len() as u64)) as usize;
+        self.slots[slot_idx].notified()
+    }
+
+    pub async fn notify<K>(&self, key: &K)
+    where
+        K: Hash,
+    {
+        let mut hasher = DefaultHasher::new();
+        key.hash(&mut hasher);
+        let hash_value = hasher.finish();
+        let slot_idx = (hash_value % (self.slots.len() as u64)) as usize;
+        self.slots[slot_idx].notify_waiters();
+    }
+
+    /// Sends the notifications in the queue.
+    // TODO: should we remove async?
```

Comment:
I think if its not required it should be removed

Reply:
removed async
justification: because it is private type in fedimint-api.